url64d634b8f0_agdturner_generic_tgz-pJ8-uk_ac_leeds_ccg_andyt_generic_visualisation_charts_Generic_BarChartJ8/src/uk/ac/leeds/ccg/andyt/generic/math/Generic_float.java

```diff
@@ -0,0 +1,325 @@
+        System.out.println(
+                "numberOfFloatsInRange(" + min + "," + max + ") "
+                + numberOfFloatsInRange);
+
+        Random a_Random = new Random(0L);
+        int n = 10;
+        //print(a_Random, n);
+        int m0 = 10;
+        for (int m = m0; m < 30; m++) {
+            System.out.println(m);
+            print(a_Random, n, m);
+        }
+
+//        BigDecimal min = new BigDecimal(Float.toString(Float.MIN_VALUE));
+//        BigDecimal value_BigDecimal = new BigDecimal(min.toString());
+//        System.out.println(value_BigDecimal.toPlainString());
+//        for (int i = 0; i < 100; i++) {
+//            value_BigDecimal = value_BigDecimal.add(min);
+//            System.out.println(value_BigDecimal.toPlainString());
+//        }
+//        System.out.println(min.toPlainString());
+//
+//        BigInteger count = BigInteger.ZERO;
+//        BigInteger divisor = new BigInteger("100000000");
+//        float value = 0;
+//        while (value < 1.0d) {
+//            value = next(value);
+//            //if (count % 100000000 == 0){
+//            if (count.remainder(divisor).compareTo(BigInteger.ZERO) == 0) {
+//                System.out.println("" + count + " values between 0 and ");
+//                System.out.println(toPlainString(value));
+//            }
+//            count = count.add(BigInteger.ONE);
+//        }
+
+//        float zero_float = 0.0d;
+//        float bigger_float = zero_float;
+//        for (int i = 0; i < 100; i++) {
+//            bigger_float = next(bigger_float);
+//            //System.out.println(bigger_float);
+//            System.out.println(toPlainString(bigger_float));
+//        }
+//
+//        float smaller_float = zero_float;
+//        for (int i = 0; i < 100; i++) {
+//            smaller_float = previous(smaller_float);
+//            //System.out.println(smaller_float);
+//            System.out.println(toPlainString(smaller_float));
+//        }
+    }
+
+    public static void print() {
+        System.out.println("Float.MIN_VALUE");
+        System.out.println(toPlainString(Float.MIN_VALUE));
+
+        System.out.println("Float.MIN_NORMAL");
+        System.out.println(toPlainString(Float.MIN_NORMAL));
+
+        System.out.println("Float.MAX_VALUE");
+        System.out.println(Float.MAX_VALUE);
+
+        //System.out.println("Float.MAX_VALUE");
+        //System.out.println(Math.nextUp());
+    }
+
+    /**
+     * Print out the next n a_Random.nextFloat() values that are represented by
+     * a string longer than m
+     * @param a_Random
+     * @param n
+     * @param m
+     */
+    public static void print(
+            Random a_Random,
+            int n,
+            int m) {
+        int counter = 0;
+        BigInteger divisor = new BigInteger("100000000");
+        BigInteger loopCounter = BigInteger.ZERO;
+        while (counter < n) {
+            float value = a_Random.nextFloat();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;a_Random&#39; (index: 0 | type: Random). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;m&#39;, &#39;n&#39;. The same parameter type is validated here for example: url64d634b8f0_agdturner_generic_tgz-pJ8-uk_ac_leeds_ccg_andyt_generic_visualisation_charts_Generic_BarChartJ8/src/uk/ac/leeds/ccg/andyt/generic/utilities/Generic_Collections.java:198. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).